src/api/java/IDecRefQueue.java

```diff
@@ -34,38 +42,31 @@ protected IDecRefQueue(int move_limit)
     {
     	m_move_limit = move_limit;
     }
- 
-    public void setLimit(int l) { m_move_limit = l; }
- 
-    protected abstract void incRef(Context ctx, long obj);
 
+    /**
+     * An implementation of this method should decrement the reference on a
+     * given native object.
+     * This function should be always called on the {@code ctx} thread.
+     *
+     * @param ctx Z3 context.
+     * @param obj Pointer to a Z3 object.
+     */
     protected abstract void decRef(Context ctx, long obj);
 
-    protected void incAndClear(Context ctx, long o)
-    {
-        incRef(ctx, o);
-        if (m_queue.size() >= m_move_limit)
-            clear(ctx);
-    }
-
-    protected void add(long o)
-    {
-        if (o == 0)
-            return;
+    public void storeReference(Context ctx, T obj) {
+        PhantomReference<T> ref = new PhantomReference<>(obj, referenceQueue);
+        referenceMap.put(ref, obj.getNativeObject());
 
-        synchronized (m_lock)
-        {
-            m_queue.add(o);
-        }
+        // TODO: use move_limit, somehow get the size of referenceQueue.
+        clear(ctx);
     }
 
     protected void clear(Context ctx)
```

Comment:
So from the way I understand the code all AST objects live as long as the context.
In other words, if an Z3Object goes out of scope, the reference to it still lingers.
Conversely, the Z3Object is invalid when the context goes out of scope.
Is there some way to get both? 
The original model appeals to situations where Z3Objects are created and disposed of en masse.
I don't understand how the new model would accommodate this.